File: python/ReconstructOriginalDigitsFromEnglish.py
"""
Given a non-empty string containing an out-of-order English representation of digits 0-9, output the digits in ascending order.

Note:
Input contains only lowercase English letters.
Input is guaranteed to be valid and can be transformed to its original digits. That means invalid inputs such as "abc" or "zerone" are not permitted.
Input length is less than 50,000.
Example 1:
Input: "owoztneoer"

Output: "012"
Example 2:
Input: "fviefuro"

Output: "45"
"""

class ReconstructOriginalDigitsFromEnglish(object):
    def originalDigits(self, s):
        """
        :type s: str
        :rtype: str
        """

        res = ""
        letters = [0]*26

        for each in s:
            letters[ord(each)-ord('a')] += 1

        #z,w,u,x,g identify unique digits
        while letters[ord('z')-ord('a')]>0:
            res += "0"
            letters[ord('z')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1
            letters[ord('r')-ord('a')] -= 1
            letters[ord('o')-ord('a')] -= 1

        while letters[ord('w')-ord('a')]>0:
            res += "2"
            letters[ord('t')-ord('a')] -= 1
            letters[ord('w')-ord('a')] -= 1
            letters[ord('o')-ord('a')] -= 1

        while letters[ord('u')-ord('a')]>0:
            res += "4"
            letters[ord('f')-ord('a')] -= 1
            letters[ord('o')-ord('a')] -= 1
            letters[ord('u')-ord('a')] -= 1
            letters[ord('r')-ord('a')] -= 1

        while letters[ord('x')-ord('a')]>0:
            res += "6"
            letters[ord('s')-ord('a')] -= 1
            letters[ord('i')-ord('a')] -= 1
            letters[ord('x')-ord('a')] -= 1

        while letters[ord('g')-ord('a')]>0:
            res += "8"
            letters[ord('e')-ord('a')] -= 1
            letters[ord('i')-ord('a')] -= 1
            letters[ord('g')-ord('a')] -= 1
            letters[ord('h')-ord('a')] -= 1
            letters[ord('t')-ord('a')] -= 1


        #now o,h,s identify unique digits
        while letters[ord('o')-ord('a')]>0:
            res += "1"
            letters[ord('o')-ord('a')] -= 1
            letters[ord('n')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1

        while letters[ord('h')-ord('a')]>0:
            res += "3"
            letters[ord('t')-ord('a')] -= 1
            letters[ord('h')-ord('a')] -= 1
            letters[ord('r')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1

        while letters[ord('s')-ord('a')]>0:
            res += "7"
            letters[ord('s')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1
            letters[ord('v')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1
            letters[ord('n')-ord('a')] -= 1


        #f and n differentiate what's left
        while letters[ord('f')-ord('a')]>0:
            res += "5"
            letters[ord('f')-ord('a')] -= 1
            letters[ord('i')-ord('a')] -= 1
            letters[ord('v')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1

        while letters[ord('n')-ord('a')]>0:
            res += "9"
            letters[ord('n')-ord('a')] -= 1
            letters[ord('i')-ord('a')] -= 1
            letters[ord('n')-ord('a')] -= 1
            letters[ord('e')-ord('a')] -= 1

        return "".join(sorted(res))


#test
tmp = ReconstructOriginalDigitsFromEnglish()
result = tmp.originalDigits("fviefuro")
print(result)

File: cpp/source/ReconstructOriginalDigitsFromEnglish.cpp

#include "../Solutions.hpp"

/************** Reconstruct Original Digits From English *****************/
/*
 Given a non-empty string containing an out-of-order English representation of digits 0-9, output the digits in ascending order.
 
 Note:
 Input contains only lowercase English letters.
 Input is guaranteed to be valid and can be transformed to its original digits. That means invalid inputs such as "abc" or "zerone" are not permitted.
 Input length is less than 50,000.
 Example 1:
 Input: "owoztneoer"
 
 Output: "012"
 Example 2:
 Input: "fviefuro"
 
 Output: "45"
 */

string Solutions::originalDigits(string s) {
    
    vector<int> digits;
    vector<int> letters(26,0);
    for (char a : s) {
        letters[a-'a']++;
    }
    
    //z,w,u,x,g identify unique digits
    if (letters['z'-'a']>0) {
        int times = letters['z'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(0);
        
        letters['z'-'a'] -= times;
        letters['e'-'a'] -= times;
        letters['r'-'a'] -= times;
        letters['o'-'a'] -= times;
    }
    if (letters['w'-'a']>0) {
        int times = letters['w'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(2);
        
        letters['t'-'a'] -= times;
        letters['w'-'a'] -= times;
        letters['o'-'a'] -= times;
    }
    if (letters['u'-'a']>0) {
        int times = letters['u'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(4);
        
        letters['f'-'a'] -= times;
        letters['o'-'a'] -= times;
        letters['u'-'a'] -= times;
        letters['r'-'a'] -= times;
    }
    if (letters['x'-'a']>0) {
        int times = letters['x'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(6);
        
        letters['s'-'a'] -= times;
        letters['i'-'a'] -= times;
        letters['x'-'a'] -= times;
    }
    if (letters['g'-'a']>0) {
        int times = letters['g'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(8);
        
        letters['e'-'a'] -= times;
        letters['i'-'a'] -= times;
        letters['g'-'a'] -= times;
        letters['h'-'a'] -= times;
        letters['t'-'a'] -= times;
    }
    
    //now o,h,s identify unique digits
    if (letters['o'-'a']>0) {
        int times = letters['o'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(1);
        
        letters['o'-'a'] -= times;
        letters['n'-'a'] -= times;
        letters['e'-'a'] -= times;
    }
    if (letters['h'-'a']>0) {
        int times = letters['h'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(3);
        
        letters['t'-'a'] -= times;
        letters['h'-'a'] -= times;
        letters['r'-'a'] -= times;
        letters['e'-'a'] -= times;
        letters['e'-'a'] -= times;
    }
    if (letters['s'-'a']>0) {
        int times = letters['s'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(7);
        letters['s'-'a'] -= times;
        letters['e'-'a'] -= times;
        letters['v'-'a'] -= times;
        letters['e'-'a'] -= times;
        letters['n'-'a'] -= times;
    }
    
    //f and n differentiate what's left
    if (letters['f'-'a']>0) {
        int times = letters['f'-'a'];
        for (int i=0;i<times;i++)
            digits.push_back(5);
        
        letters['f'-'a'] -= times;
        letters['i'-'a'] -= times;
        letters['v'-'a'] -= times;
        letters['e'-'a'] -= times;
    }
    if (letters['n'-'a']>0) {
        int times = letters['n'-'a'];
        times >>= 1;//nine has two 'n'
        for (int i=0;i<times;i++)
            digits.push_back(9);
        
        letters['n'-'a'] -= times;
        letters['i'-'a'] -= times;
        letters['n'-'a'] -= times;
        letters['e'-'a'] -= times;
    }
    
    sort(digits.begin(),digits.end());
    string res;
    for (int i:digits) {
        res += to_string(i);
    }
    return res;
}
